### PR TITLE
[SDL2] cmake: add support for targeting QNX

### LINKED ISSUE #8321: Cross-compilation of SDL2 failed for QNX
I am cross compiling SDL2 for Qualcomm QNX on Ubuntu Focal, which is failing with the following error at CMake generate step:

```
CMake Error at cmake/macros.cmake:36 (message):
  *** ERROR: Threads are needed by many SDL subsystems and may not be
  disabled
Call Stack (most recent call first):
  CMakeLists.txt:2965 (message_error)
```

I used SDL2 branch (commit hash: `6b5eadb10f02b51b400e710e2c95e9f0c2cb579c`) of SDL repo and this is the CMake toolchain file `aarch64-qnx.cmake`:

```
set(CMAKE_SYSROOT $ENV{QNX_TARGET})
set(CMAKE_SYSTEM_NAME QNX)
set(CMAKE_SYSTEM_VERSION 710)
set(CMAKE_SYSTEM_PROCESSOR "aarch64")
set(arch nto${CMAKE_SYSTEM_PROCESSOR})

set(CMAKE_AS "$ENV{QNX_HOST}/usr/bin/${arch}-as")
set(CMAKE_AR "$ENV{QNX_HOST}/usr/bin/${arch}-ar")
set(CMAKE_C_COMPILER "$ENV{QNX_HOST}/usr/bin/${arch}-gcc")
set(CMAKE_CXX_COMPILER "$ENV{QNX_HOST}/usr/bin/${arch}-g++")
set(CMAKE_LINKER "$ENV{QNX_HOST}/usr/bin/${arch}-ld")

set(CMAKE_C_COMPILER_TARGET "gcc_ntoaarch64le")
set(CMAKE_CXX_COMPILER_TARGET "gcc_ntoaarch64le")
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fexceptions -fPIC -D_XOPEN_SOURCE=600")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fexceptions -fPIC -D_XOPEN_SOURCE=600 -std=gnu++14")

add_definitions(-D_POSIX_C_SOURCE=200112L -D_QNX_SOURCE)

set(CMAKE_FIND_LIBRARY_PREFIXES "lib")
set(CMAKE_FIND_LIBRARY_SUFFIXES ".so;.a")

set(CMAKE_FIND_ROOT_PATH "${CMAKE_SYSROOT}")
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
```

The CMake version is `3.26.1`. 

CMake generate command:

`cmake -B build -DCMAKE_TOOLCHAIN_FILE=$(pwd)/aarch64-qnx.cmake -DSDL_X11=0 -DCMAKE_BUILD_TYPE=Release`

Any help would be much appreciated. Thanks.

### PR DESCRIPTION
**Pull request by madebr**
Backport of #7018 and #7041

@elahav Can you review these changes?

## Existing Issue(s)
Fixes #8321

### CODE REVIEW

**Comment by elahav** [comment]
Looks good. I assume this version still has the QNX video and audio support that I had to put back to SDL3?
I can try building and testing this over the weekend.

File: cmake/sdlplatform.cmake [hunk 1]
```diff
@@ -42,6 +42,8 @@ macro(SDL_DetectCMakePlatform)
         set(SDL_CMAKE_PLATFORM IOS)
       endif()
     endif()
+  elseif(QNX)
```
Comment (elahav) [inline cmake/sdlplatform.cmake:45]: This does not detect QNX properly. First, `QNX` is not yet defined. Second, it should be under the `UNIX` section above.

```
diff --git a/cmake/sdlplatform.cmake b/cmake/sdlplatform.cmake
index 536a3fb2b..6d8ece6ac 100644
--- a/cmake/sdlplatform.cmake
+++ b/cmake/sdlplatform.cmake
@@ -28,6 +28,8 @@ macro(SDL_DetectCMakePlatform)
       set(SDL_CMAKE_PLATFORM AIX)
     elseif(CMAKE_SYSTEM_NAME MATCHES "Minix.*")
       set(SDL_CMAKE_PLATFORM MINIX)
+    elseif(CMAKE_SYSTEM_NAME MATCHES "QNX")
+      set(SDL_CMAKE_PLATFORM QNX)
     endif()
   elseif(APPLE)
     if(CMAKE_SYSTEM_NAME MATCHES ".*Darwin.*")
@@ -42,8 +44,6 @@ macro(SDL_DetectCMakePlatform)
         set(SDL_CMAKE_PLATFORM IOS)
       endif()
     endif()
-  elseif(QNX)
-    set(SDL_CMAKE_PLATFORM QNX)
   elseif(CMAKE_SYSTEM_NAME MATCHES "BeOS.*")
     message_error("BeOS support has been removed as of SDL 2.0.2.")
   elseif(CMAKE_SYSTEM_NAME MATCHES "Haiku.*")
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -1456,6 +1458,12 @@ elseif(UNIX AND NOT APPLE AND NOT RISCOS AND NOT HAIKU)
         file(GLOB AIX_AUDIO_SOURCES ${SDL2_SOURCE_DIR}/src/audio/paudio/*.c)
         list(APPEND SOURCE_FILES ${AIX_AUDIO_SOURCES})
         set(HAVE_SDL_AUDIO TRUE)
+    elseif(QNX)
+        set(SDL_AUDIO_DRIVER_QSA 1)
+        file(GLOB QSA_AUDIO_SOURCES ${SDL3_SOURCE_DIR}/src/audio/qsa/*.c)
+        list(APPEND SOURCE_FILES ${SDL_AUDIO_DRIVER_QSA})
```
Comment (elahav) [inline CMakeLists.txt:1464]: Typo here results in an odd cmake failure:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 60f79b5e7..93a46ebe3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1461,7 +1461,7 @@ elseif(UNIX AND NOT APPLE AND NOT RISCOS AND NOT HAIKU)
     elseif(QNX)
         set(SDL_AUDIO_DRIVER_QSA 1)
         file(GLOB QSA_AUDIO_SOURCES ${SDL3_SOURCE_DIR}/src/audio/qsa/*.c)
-        list(APPEND SOURCE_FILES ${SDL_AUDIO_DRIVER_QSA})
+        list(APPEND SOURCE_FILES ${QSA_AUDIO_SOURCES})
         list(APPEND EXTRA_LIBS asound)
         set(HAVE_SDL_AUDIO TRUE)
     endif()
```

**Comment by elahav** [comment]
I have a problem when cross-compiling in that pkg-config finds Linux host packages and declares support for various subsystems (e.g., PulseAudio, KMSDRM). Is there a common solution to this problem, or do I need to exclude all these options explicitly?

**Comment by sezero** [comment]
If qnx configuration has nothing that relies on pkg-config discovery, then you can set `PKG_CONFIG=/bin/false` environment. Otherwise, explicit exclusion looks like the way to me, as far as I can see.

**Comment by elahav** [comment]
@sezero That results in cmake failures:

```
CMake Warning at cmake/sdlchecks.cmake:128 (message):
  Unable to found the alsa development library
Call Stack (most recent call first):
  CMakeLists.txt:1469 (CheckALSA)


CMake Error at cmake/sdlchecks.cmake:198 (pkg_check_modules):
  Unknown CMake command "pkg_check_modules".
Call Stack (most recent call first):
  CMakeLists.txt:1470 (CheckJACK)
```

Is this not a common problem when cross-compiling for other systems? I can build on a QNX system, but that's not an option for the vast majority of users.

**Comment by sezero** [comment]
> @sezero That results in cmake failures:

As I said: `If qnx configuration has nothing that relies on pkg-config discovery, then ` :smile::smile: 

> Is this not a common problem when cross-compiling for other systems?

It can be, yes. One solution is building and including a cross-targeted pkg-config in the
cross-toolchain, like x86_64-xxx-qnx-pkgconfig or whatever and setting PKG_CONFIG
environment to that file.

Another solution in our cmake and autotools scripts would be checking whether linkage
actually succeeds after pkg-config thinks that it found something.

**Comment by sezero** [comment]
> To be clear, neither ALSA nor JACK should be used for the QNX build. I was hoping that CMake not being able to use pkg-config would result in it excluding these modules and the build would continue without.

Hm, I just tested native configuration on my linux with `PKG_CONFIG=/bin/false` set, like:

```sh
$ mkdir b && cd b
$ PKG_CONFIG=/bin/false cmakerun.sh cmake -DCMAKE_BUILD_TYPE=Release ..
```

The only warning I get is this: (as extracted from terminal output)
```
-- Detecting CXX compile features - done
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
    Reason given by package: The command
      "/bin/false" --version
    failed with output
      

-- Performing Test CHECK_CPU_ARCHITECTURE_X64
```

.. and every pkgconfig-dependent checks seem to be omitted.

This is with cmake-3.18.6. What is your cmake version, maybe it's related?

**Comment by elahav** [comment]
```
elahav@lamia:~$ cmake --version
cmake version 3.22.1
```

**Comment by sezero** [comment]
> ```
> elahav@lamia:~$ cmake --version
> cmake version 3.22.1
> ```

OK, maybe @madebr has an insight for this.

**Comment by sezero** [comment]
P.S.: You know you can use autotools for this with PKG_CONFIG env varibable set, yes?

**Comment by madebr** [comment]
There are a few ways to do what you want:
- Disable `pkg-config` entirely by configuring with `-DPKG_CONFIG_EXECUTABLE=/bin/false` or environment variable `PKG_CONFIG`.
Important to know is that this only works reliably when you configure the first time: CMake will cache the name of the `pkg-config` executable and the variables it returns.
- Disable `pkg-config` entirely by configuring with [`-DCMAKE_DISABLE_FIND_PACKAGE_PkgConfig=TRUE`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html). This does (currently) not work with SDL.
- Configure with the environment variable `PKG_CONFIG_LIBDIR` and `PKG_CONFIG_PATH` set to the pkgconfig dir of your target (or to a dummy directory).
 I think this latest option is the best. It can also be done in a cmake toolchain file by doing:
    ```cmake
    set(ENV{PKG_CONFIG_LIBDIR} "/path/to/qnx/toolchain/sys-root/lib/pkgconfig")
    set(ENV{PKG_CONFIG_PATH} "/path/to/qnx/toolchain/sys-root/lib/pkgconfig")
    ```

File: cmake/sdlplatform.cmake [hunk 3]
```diff
@@ -29,6 +29,8 @@ macro(SDL_DetectCMakePlatform)
     elseif(CMAKE_SYSTEM_NAME MATCHES "Minix.*")
       set(SDL_CMAKE_PLATFORM MINIX)
     endif()
+  elseif(CMAKE_SYSTEM_NAME MATCHES "QNX")
```
Comment (elahav) [inline cmake/sdlplatform.cmake:32]: This is still outside the UNIX section.

```
diff --git a/cmake/sdlplatform.cmake b/cmake/sdlplatform.cmake
index 5659dfb30..6d8ece6ac 100644
--- a/cmake/sdlplatform.cmake
+++ b/cmake/sdlplatform.cmake
@@ -28,9 +28,9 @@ macro(SDL_DetectCMakePlatform)
       set(SDL_CMAKE_PLATFORM AIX)
     elseif(CMAKE_SYSTEM_NAME MATCHES "Minix.*")
       set(SDL_CMAKE_PLATFORM MINIX)
+    elseif(CMAKE_SYSTEM_NAME MATCHES "QNX")
+      set(SDL_CMAKE_PLATFORM QNX)
     endif()
-  elseif(CMAKE_SYSTEM_NAME MATCHES "QNX")
-    set(SDL_CMAKE_PLATFORM QNX)
   elseif(APPLE)
     if(CMAKE_SYSTEM_NAME MATCHES ".*Darwin.*")
       set(SDL_CMAKE_PLATFORM DARWIN)
```

Comment (madebr) [reply cmake/sdlplatform.cmake]: Oops. Dumb me. Thanks for double checking this.

**Comment by elahav** [comment]
> I think this latest option is the best. It can also be done in a cmake toolchain file by doing:

Yes, that works. I even used your fake paths as-is in the toolchain file ;-)

**Comment by elahav** [comment]
The files under `src/video/qnx` and `src/audio/qsa` are not being built. I'm looking into it.

File: cmake/sdlchecks.cmake [hunk 4]
```diff
@@ -911,6 +911,22 @@ macro(CheckOpenGLES)
...
+# Requires:
+# - EGL
+macro(CheckQNXScreen)
+  if(QNX AND HAVE_OPENGL_EGL)
+    check_c_source_compiles("
+        #include <screen/screen.h>
+        int main (int argc, char** argv) { return 0; }" HAVE_QNX_SCREEN)
+    if(HAVE_QNX_SCREEN)
+      set(SDL_VIDEO_DRIVER_QNX 1)
+      file(GLOB QNX_VIDEO_SOURCES ${SDL3_SOURCE_DIR}/src/video/qnx/*.c)
```
Comment (elahav) [inline cmake/sdlchecks.cmake:923]: Typo: SDL3 => SDL2

File: CMakeLists.txt [hunk 5]
```diff
@@ -1456,6 +1458,12 @@ elseif(UNIX AND NOT APPLE AND NOT RISCOS AND NOT HAIKU)
         file(GLOB AIX_AUDIO_SOURCES ${SDL2_SOURCE_DIR}/src/audio/paudio/*.c)
         list(APPEND SOURCE_FILES ${AIX_AUDIO_SOURCES})
         set(HAVE_SDL_AUDIO TRUE)
+    elseif(QNX)
+        set(SDL_AUDIO_DRIVER_QSA 1)
+        file(GLOB QSA_AUDIO_SOURCES ${SDL3_SOURCE_DIR}/src/audio/qsa/*.c)
```
Comment (elahav) [inline CMakeLists.txt:1463]: Typo: SDL3 => SDL2

File: cmake/sdlchecks.cmake (hunk 4 above)
Comment (madebr) [reply cmake/sdlchecks.cmake]: Thanks!

**Comment by elahav** [comment]
Needs another patch:

```
diff --git a/src/audio/qsa/SDL_qsa_audio.c b/src/audio/qsa/SDL_qsa_audio.c
index 0ab477485..1a80f94f0 100644
--- a/src/audio/qsa/SDL_qsa_audio.c
+++ b/src/audio/qsa/SDL_qsa_audio.c
@@ -233,6 +233,7 @@ static Uint8 *QSA_GetDeviceBuf(_THIS)
 static void QSA_CloseDevice(_THIS)
 {
     if (this->hidden->audio_handle != NULL) {
+#if _NTO_VERSION < 710
         if (!this->iscapture) {
             /* Finish playing available samples */
             snd_pcm_plugin_flush(this->hidden->audio_handle,
@@ -242,6 +243,7 @@ static void QSA_CloseDevice(_THIS)
             snd_pcm_plugin_flush(this->hidden->audio_handle,
                                  SND_PCM_CHANNEL_CAPTURE);
         }
+#endif
         snd_pcm_close(this->hidden->audio_handle);
     }
```

`snd_pcm_plugin_flush` was removed in QNX 7.1 and later.

With all of the reported issues corrected I can build SDL2

**Comment by elahav** [comment]
...and Prince of Persia works on RaspberryPi 4, so I guess we are good!

**Comment by elahav** [comment]
Verified that the latest version builds and works. For the record, this is my toolchain file:

```
set(CMAKE_SYSTEM_NAME QNX)

set(arch gcc_ntoaarch64le)

set(CMAKE_C_COMPILER qcc)
set(CMAKE_C_COMPILER_TARGET ${arch})
set(CMAKE_CXX_COMPILER q++)
set(CMAKE_CXX_COMPILER_TARGET ${arch})

set(CMAKE_SYSROOT $ENV{QNX_TARGET})

set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)

set(ENV{PKG_CONFIG_LIBDIR} "/do_not_use_pkg_config_when_cross_compiling")
set(ENV{PKG_CONFIG_PATH} "/do_not_use_pkg_config_when_cross_compiling")
```

**Comment by madebr** [comment]
> ...and Prince of Persia works on RaspberryPi 4, so I guess we are good!

Awesome! Thanks for verifying it!

**Pull request merged**